ZenPacks/zenoss/OpenStackInfrastructure/datasources/ApiEndpointStatusDataSource.py

```diff
@@ -0,0 +1,365 @@
+            'username': context.zCommandUsername,
+            'password': context.zCommandPassword,
+            'project_id': context.zOpenStackProjectId,
+            'auth_url': context.zOpenStackAuthUrl,
+            'auth_required': datasource.auth_required,
+            'ok_result': datasource.ok_result
+        }
+
+    def _get_ds0(self, config):
+        # the monitoring template for this datasource contains multiple
+        # datasources, one for each supported service type.
+        #
+        # For the component being monitored (there will be only one in the
+        # config for each invocation of collect, because the component id is
+        # part of the config_key above), find the datasource whose name
+        # matches the modeled service type for this api endpoint.
+
+        datasources = [x for x in config.datasources
+                       if x.datasource == x.params['service_type']]
+        if datasources:
+            return datasources[0]
+
+    @inlineCallbacks
+    def collect(self, config):
+        ds0 = self._get_ds0(config)
+        if not ds0:
+            returnValue(None)
+
+        service_type = ds0.params['service_type']
+        url = ds0.params['url']
+
+        log.debug("Collect for OpenStack API Endpoint Status (%s / %s  (%s endpoint))" % (config.id, url, service_type))
+
+        client = APIClient(
+            ds0.params['username'],
+            ds0.params['password'],
+            ds0.params['auth_url'],
+            ds0.params['project_id']
+        )
+
+        if ds0.params['auth_required']:
+            result = yield client.authenticated_get(url)
+        else:
+            result = yield client.get(url)
+        returnValue(result)
+
+    def onSuccess(self, result, config):
+        data = self.new_data()
+        ds0 = self._get_ds0(config)
+        if not ds0:
+            return data
+
+        summary = '%s API Status: successful collection' % ds0.params['service_type'].title()
+        data['events'].append({
+            'device': config.id,
+            'component': ds0.component,
+            'summary': summary,
+            'severity': ZenEventClasses.Clear,
+            'eventClassKey': 'openStackApiEndpointStatus',
+            'eventKey': ds0.params['url']
+        })
+
+        return data
+
+    def onError(self, result, config):
+        data = self.new_data()
+        ds0 = self._get_ds0(config)
+        if not ds0:
+            return data
+
+        summary = '%s API Status: %s' % (ds0.params['service_type'].title(), result_errmsg(result))
+        data['events'].append({
+            'device': config.id,
+            'component': ds0.component,
+            'summary': summary,
+            'severity': ZenEventClasses.Error,
+            'eventClassKey': 'openStackApiEndpointStatus',
+            'eventKey': ds0.params['url']
+        })
+
+        log.error('%s: %s', config.id, data['events'][-1]['summary'])
+
+        return data
+
+
+@inlineCallbacks
```

Comment:
Yes, it's just for testing / figuring out the settings I put in the monitoring template.  I don't really anticipate needing to use it again, but just wanted to save it.   It could go in a separate file, but this is how it was done in txapiclient.py and I figured it was probably fine for our purposes.